core/src/main/scala/kafka/server/BrokerLifecycleManager.scala

```diff
@@ -453,79 +490,73 @@ class BrokerLifecycleManager(
           scheduleNextCommunicationAfterFailure()
         }
       }
     }
-
-    override def onTimeout(): Unit = {
-      info("Unable to send a heartbeat because the RPC got timed out before it could be sent.")
-      scheduleNextCommunicationAfterFailure()
-    }
   }
 
-  private class BrokerHeartbeatResponseEvent(response: BrokerHeartbeatResponseData) extends EventQueue.Event {
-    override def run(): Unit = {
-      failedAttempts = 0
-      _state match {
-        case BrokerState.STARTING =>
-          if (response.isCaughtUp) {
-            info(s"The broker has caught up. Transitioning from STARTING to RECOVERY.")
-            _state = BrokerState.RECOVERY
-            initialCatchUpFuture.complete(null)
-          } else {
-            debug(s"The broker is STARTING. Still waiting to catch up with cluster metadata.")
-          }
-          // Schedule the heartbeat after only 10 ms so that in the case where
-          // there is no recovery work to be done, we start up a bit quicker.
-          scheduleNextCommunication(NANOSECONDS.convert(10, MILLISECONDS))
-        case BrokerState.RECOVERY =>
-          if (!response.isFenced) {
-            info(s"The broker has been unfenced. Transitioning from RECOVERY to RUNNING.")
-            initialUnfenceFuture.complete(null)
-            _state = BrokerState.RUNNING
-          } else {
-            info(s"The broker is in RECOVERY.")
-          }
-          scheduleNextCommunicationAfterSuccess()
-        case BrokerState.RUNNING =>
-          debug(s"The broker is RUNNING. Processing heartbeat response.")
-          scheduleNextCommunicationAfterSuccess()
-        case BrokerState.PENDING_CONTROLLED_SHUTDOWN =>
-          if (!response.shouldShutDown()) {
-            info(s"The broker is in PENDING_CONTROLLED_SHUTDOWN state, still waiting " +
-              "for the active controller.")
-            if (!gotControlledShutdownResponse) {
-              // If this is the first pending controlled shutdown response we got,
-              // schedule our next heartbeat a little bit sooner than we usually would.
-              // In the case where controlled shutdown completes quickly, this will
-              // speed things up a little bit.
-              scheduleNextCommunication(NANOSECONDS.convert(50, MILLISECONDS))
-            } else {
-              scheduleNextCommunicationAfterSuccess()
-            }
-          } else {
-            info(s"The controller has asked us to exit controlled shutdown.")
-            beginShutdown()
-          }
-          gotControlledShutdownResponse = true
-        case BrokerState.SHUTTING_DOWN =>
-          info(s"The broker is SHUTTING_DOWN. Ignoring heartbeat response.")
-        case _ =>
-          error(s"Unexpected broker state ${_state}")
-          scheduleNextCommunicationAfterSuccess()
-      }
+  private def scheduleNextCommunicationImmediately(): Unit = {
+    if (communicationInFlight) {
```

Comment:
I am not sure if this completely avoids duplicated `HeartbeatRequest`s. Consider the following flow. The event queue contains a `BrokerHeartbeatResponseEvent` followed by an `OfflineDirEvent`. The event queue thread processes the former.  It sets `communicationInFlight` to false and enqueues a `CommunicationEvent`. It then process the `OfflineDirEvent`. Since `communicationInFlight` is false, it enqueues another `CommunicationEvent`. Now, there are two `CommunicationEvent`s and each will introduce a separate `HeartbeatRequest`.

Reply:
You are correct, thanks for finding this. I placed the `communicationInFlight` check in `scheduleNextCommunicationImmediately`, but `BrokerHeartbeatResponseEvent` bypasses that by calling `scheduleNextCommunication` directly.

I'm moving the check to `scheduleNextCommunication` so we can catch all cases. Since `CommunicationEvent` is only scheduled from this function, we can make sure to not schedule it when a request is in flight.